web3.js/src/publickey.ts

```diff
@@ -54,22 +61,41 @@ export class PublicKey extends Struct {
   constructor(value: PublicKeyInitData) {
     super({});
     if (isPublicKeyData(value)) {
-      this._bn = value._bn;
+      if (typeof value._bn === 'object') {
+        // Legacy implementation of public key storage as a `BN` instance.
+        this._bn = toBigIntLE(value._bn.toBuffer('le', 32));
```

Comment:
I _think_ this is OK, actually.

* Someone supplies a `BN` instance to the constructor (endianness is irrelevant at that point)
* We convert it to an LE buffer
* We convert that LE buffer to a `bigint` using `toBigIntLE` (which implements ‘convert this LE buffer to a `bigint`’)

Now, the internal representation is a `bigint`, which doesn't have an endianness – it just exists.

---

The way in which this _is_ a breaking change is for anyone who ever reaches into `PublicKey::_bn`. We never marked that TypeScript `private` so there has never been and still is no compile time protection against someone doing so. If we ship this, they will be surprised to find that `_bn` is no longer a `BN` but a `bigint`, and their code will break.

Reply:
Oh duh, of course, you're right, sorry :facepalm: I got confused with using borsh encoding that just treated the underlying as little-endian always, which isn't an issue here because you're doing `toBuffer('le'`

Reply:
> If we ship this, they will be surprised to find that `_bn` is no longer a `BN` but a `bigint`

What if we change the name and get rid of the internal `_bn` property, or throw when it's accessed with a getter? This will cause access of internals to fail faster, rather than experience a larger range of errors where they get the value and pass it to something else or try to call a method on it that doesn't exist.

Reply:
Yeah, we could do that. We'll have to _add_ something to the constructor which allows `PublicKey` to be part of the `PublicKeyInitData` union, because people have gotten used to doing this:

```
// Like, I've actually seen this.
new PublicKey(new PublicKey(...));
```

Right now that _works_ because `PublicKey` just _happens_ to conform to `{_bn: BN}` because, damnit, `_bn` was never marked `private`.